java/src/main/native/OrtJniUtil.c

```diff
@@ -600,427 +588,473 @@ size_t copyPrimitiveArrayToJava(JNIEnv *jniEnv, ONNXTensorElementDataType onnxTy
         case ONNX_TENSOR_ELEMENT_DATA_TYPE_UNDEFINED:
         default: {
             throwOrtException(jniEnv, convertErrorCode(ORT_NOT_IMPLEMENTED), "Invalid tensor element type.");
-            return 0;
+            return -1;
         }
     }
 }
 
-size_t copyTensorToJava(JNIEnv *jniEnv, ONNXTensorElementDataType onnxType, uint8_t* tensor, size_t tensorSize,
+int64_t copyTensorToJava(JNIEnv *jniEnv, ONNXTensorElementDataType onnxType, uint8_t* tensor, size_t tensorSize,
                         size_t dimensionsRemaining, jarray output) {
-    if (dimensionsRemaining == 1) {
-        // write out 1d array of the respective primitive type
-        return copyPrimitiveArrayToJava(jniEnv,onnxType,tensor,output);
-    } else {
-        // recurse through the dimensions
-        // Java arrays are objects until the final dimension
-        jobjectArray outputObjArr = (jobjectArray) output;
-        uint32_t dimLength = (*jniEnv)->GetArrayLength(jniEnv,outputObjArr);
-        size_t sizeConsumed = 0;
-        for (uint32_t i = 0; i < dimLength; i++) {
-            jarray childArr = (jarray) (*jniEnv)->GetObjectArrayElement(jniEnv,outputObjArr,i);
-            sizeConsumed += copyTensorToJava(jniEnv, onnxType, tensor + sizeConsumed, tensorSize - sizeConsumed, dimensionsRemaining - 1, childArr);
-            // Cleanup reference to childArr so it doesn't prevent GC.
-            (*jniEnv)->DeleteLocalRef(jniEnv,childArr);
-        }
-        return sizeConsumed;
+  if (dimensionsRemaining == 1) {
+    // write out 1d array of the respective primitive type
+    return copyPrimitiveArrayToJava(jniEnv, onnxType, tensor, output);
+  } else {
+    // recurse through the dimensions
+    // Java arrays are objects until the final dimension
+    jobjectArray outputObjArr = (jobjectArray) output;
+    int32_t dimLength = (*jniEnv)->GetArrayLength(jniEnv, outputObjArr);
+    int64_t sizeConsumed = 0;
+    for (int32_t i = 0; i < dimLength; i++) {
+      jarray childArr = (jarray) (*jniEnv)->GetObjectArrayElement(jniEnv, outputObjArr, i);
+      int64_t consumed = copyTensorToJava(jniEnv, onnxType, tensor + sizeConsumed, tensorSize - sizeConsumed, dimensionsRemaining - 1, childArr);
+      sizeConsumed += consumed;
+      // Cleanup reference to childArr so it doesn't prevent GC.
+      (*jniEnv)->DeleteLocalRef(jniEnv, childArr);
+      // If we failed to copy an array then break and return.
+      if (consumed == -1) {
+        return -1;
+      }
```

Comment:
Should we throw here?

Reply:
Same as above, `copyPrimitiveArrayToJava` has already thrown if we have a -1 here.